### PR TITLE
[Web panel] Loaded tab contents in web panel

### LINKED ISSUE #49794: Load new pinned tab's contents to panel view and make it persisted

### PR DESCRIPTION
**Pull request by simonhong**
Resolves https://github.com/brave/brave-browser/issues/49794

Security review: https://github.com/brave/reviews/issues/2113

We ues tab contents for web panel to make web panel contents have same security level with tab contents.
And `BraveMultiContentsView` provides another space for showing panel contents.

When web panel item is clicked, `SidebarWebPanelController::OpenWebPanel()`
creates pinned tab and set that tab's contents as web panel's contents.
`BraveMultiContentsView::contents_container_view_for_web_panel_` works as its holder. 
Whenever its active stage changes, its focus border is also updated.
For now, same focus border is used with split view.

As we can have split view and web panel both at once, this could conflict with upstream's code.
Upstream assumes that  current active tab is split tab when split view is opened. But it couldn't with web panel.
To avoid that conflict, `BraveMultiContentsViewDelegateImpl` is introduced and
`BraveBrowserView::IsWebPanelContents()` is used to skip that conflict.

TEST=SidebarBrowserWithWebPanelTest.WebPanelTest

Manual test:
1. Launch Brave with ` --enable-features=SidebarWebPanel`
2. Add some sidebar items
3. Click that item and check panel is opened.
4. See below recordings for more testing scenarios

https://github.com/user-attachments/assets/ffdad8de-5a9e-4cbc-9716-b5d0fd2bac1b

### CODE REVIEW

**Comment by zeusshomaly** [comment]
You are doing gods work mate. I am so happy! Thank you so much! (I won't try to annoy you, apologises if it seems that way, I just cannot thank you enough)

**Comment by zeusshomaly** [comment]
And no offence, what's the need of this? The G to the left of the current tab you're on? in Edge it doesn't do that? or this is just a draft and won't happen when the exact web panel comes out?
Regards

image

**Comment by zeusshomaly** [comment]
Very proud that the web panel works with split view now all at the same time, thanks Simon! May I kindly ask, I really hope the tab (seen there) on your tabs won't appear when you open any app in the sidebar web panel. This is very odd to look when it's out in the released version (when web panel is fully finished; take your time) seen below in purple (circled):
image

Regards.

**Comment by simonhong** [comment]
> Very proud that the web panel works with split view now all at the same time, thanks Simon! May I kindly ask, I really hope the tab (seen there) on your tabs won't appear when you open any app in the sidebar web panel. This is very odd to look when it's out in the released version (when web panel is fully finished; take your time) seen below in purple (circled): image
> 
> Regards.

Wow, you already saw my recordings! As we loading side panel's contents in a tab model, another tab is created as a pinned tab type. We also want to hide it. We'll try to hide :) but not in this PR.

**Comment by zeusshomaly** [comment]
> > Very proud that the web panel works with split view now all at the same time, thanks Simon! May I kindly ask, I really hope the tab (seen there) on your tabs won't appear when you open any app in the sidebar web panel. This is very odd to look when it's out in the released version (when web panel is fully finished; take your time) seen below in purple (circled): image
> > Regards.
> 
> Wow, you already saw my recordings! As we loading side panel's contents in a tab model, another tab is created as a pinned tab type. We also want to hide it. We'll try to hide :) but not in this PR.

Haha I am fast mate. I am very excited for web panel. And thank you, it does look weird with it haha.

Also, before it comes out to nightly (web panel) are you able to make some sort of UI like that to show on the sidebar app icon of whatever is playing audio so you can hover over it to pause it? it'd be pretty cool! Pause, next, go back when minimised even :)
image

Edge's replica:
image


And simon while i'm at it,
I also see that on Web Panel's To Do List it shows 'Pinned tab for web panel should not be syncable'. That is fine, but here (https://github.com/brave/brave-browser/issues/49641) you can see that Aguscruiz wants Sidebar bookmarks syncable. So I am a bit confused there.
Edit: no hurry at all, but it does seem confusing tho, and I think pinned tab and sidebar bookmark are both the same thing?
image

I do think with the syncable thing you should have a word with agus if you can :)

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 1]
```diff
@@ -116,6 +131,19 @@ void BraveMultiContentsView::ResetResizeArea() {
...
 }
 
+void BraveMultiContentsView::UpdateContentsBorderAndOverlay() {
+  if (!contents_container_view_for_web_panel_ ||
+      !contents_container_view_for_web_panel_->is_active()) {
+    MultiContentsView::UpdateContentsBorderAndOverlay();
+    return;
+  }
+
+  for (auto* contents_container_view : contents_container_views_) {
```
Comment (simonhong) [inline browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: Add comment - Set both split tab looks inactive as web panel is active.

File: browser/ui/views/frame/brave_browser_view.cc [hunk 2]
```diff
@@ -813,6 +820,20 @@ bool BraveBrowserView::MaybeUpdateDevtools(content::WebContents* web_contents) {
   return result;
 }
 
+bool BraveBrowserView::MaybeUpdateSplitView(
+    content::WebContents* web_contents) {
+  if (!multi_contents_view_) {
+    return BrowserView::MaybeUpdateSplitView(web_contents);
+  }
+
+  if (IsWebPanelContents(web_contents) &&
```
Comment (simonhong) [inline browser/ui/views/frame/brave_browser_view.cc]: Add comment here.

File: chromium_src/chrome/browser/ui/views/frame/browser_view.cc [hunk 3]
```diff
@@ -32,14 +33,22 @@
...
 
 #define BRAVE_BROWSER_VIEW_LAYOUT_CONVERTED_HIT_TEST \
   if (dst->GetWidget() != src->GetWidget()) {        \
     return false;                                    \
   }
 
+// When web panel's tab is activated, don't need to change
+// tab contents. Tab contents is still same.
+#define BRAVE_BROWSER_VIEW_ON_ACTIVE_TAB_CHANGED \
+  change_tab_contents &= !IsWebPanelContents(new_contents);
```
Comment (sangwoo108) [inline chromium_src/chrome/browser/ui/views/frame/browser_view.cc]: I guess this line also should covers where

```
old_contents == web_panel_contents &&
(new_contents == multi_contents_view_->GetActiveContentsView()->web_contents() ||
 new_contents == multi_contents_view_->GetInactiveContentsView()->web_contents())
```

i.e. active tab is changed from web panel contents to one of visible tabs in contents area.

File: browser/ui/views/frame/brave_browser_view.cc [hunk 4]
```diff
@@ -812,6 +819,26 @@ bool BraveBrowserView::MaybeUpdateDevtools(content::WebContents* web_contents) {
...
+  if (!multi_contents_view_) {
+    return BrowserView::MaybeUpdateSplitView(web_contents);
+  }
+
+  // Don't need to update split view state if |web_contents| is web panel.
+  // In BrowserView::MaybeUpdateSplitView(), there is assumption that
+  // split tab is active when multi contents view shows split view now.
+  // But, it's not when web panel is active and split view is opened together.
+  // Early return to avoid crash from that. If |web_contents| is not related
+  // with split tab, dont' need to call base class' method.
```
Comment (sangwoo108) [inline browser/ui/views/frame/brave_browser_view.cc]:
```suggestion
  // with split tab, don't need to call base class' method.
```

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 5]
```diff
@@ -54,11 +54,12 @@ void BraveMultiContentsView::UseContentsContainerViewForWebPanel() {
...
-void BraveMultiContentsView::SetWebPanelVisible(bool visible) {
+void BraveMultiContentsView::SetWebPanelContents(
+    content::WebContents* web_contents) {
   CHECK(contents_container_view_for_web_panel_);
-  contents_container_view_for_web_panel_->SetVisible(visible);
-  contents_container_view_for_web_panel_->UpdateBorderAndOverlay(true, true,
-                                                                 false);
+  contents_container_view_for_web_panel_->contents_view()->SetWebContents(
+      web_contents);
+  contents_container_view_for_web_panel_->SetVisible(!!web_contents);
```
Comment (sangwoo108) [inline browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: suggestion: no need to explicitly cast contents to boolean?

```suggestion
  contents_container_view_for_web_panel_->SetVisible(web_contents);
```

File: browser/ui/sidebar/sidebar_web_panel_controller.h [hunk 6]
```diff
@@ -7,32 +7,45 @@
...
   SidebarWebPanelController(const SidebarWebPanelController&) = delete;
   SidebarWebPanelController& operator=(const SidebarWebPanelController&) =
       delete;
 
-  void OpenWebPanel(const SidebarItem& item);
-  void CloseWebPanel();
-  bool IsShowingWebPanel() const;
+  void ToggleWebPanel(const SidebarItem& item);
+
+  content::WebContents* panel_contents() { return panel_contents_.get(); }
```
Comment (sangwoo108) [inline browser/ui/sidebar/sidebar_web_panel_controller.h]: Can this be const? I see only one place that uses this and it compares pointer.


```suggestion
  const content::WebContents* panel_contents() const { return panel_contents_.get(); }
```

File: browser/ui/sidebar/sidebar_web_panel_controller.cc [hunk 7]
```diff
@@ -7,23 +7,48 @@
...
 
 SidebarWebPanelController::SidebarWebPanelController(BrowserView& browser_view)
     : browser_view_(browser_view) {
   CHECK(IsWebPanelFeatureEnabled());
+  browser_view_->browser()->tab_strip_model()->AddObserver(this);
 }
 
-SidebarWebPanelController::~SidebarWebPanelController() = default;
+SidebarWebPanelController::~SidebarWebPanelController() {
+  CHECK(!panel_contents_);
```
Comment (sangwoo108) [inline browser/ui/sidebar/sidebar_web_panel_controller.cc:27]: I'd be worth having comment like

```
When browser closes while panel is open, tab strip model will be cleared and panel_contents_ is also will be cleared out from SidebarWebPanelController::OnTabRemoved()
```

File: browser/ui/sidebar/sidebar_web_panel_controller.cc [hunk 8]
```diff
@@ -7,23 +7,48 @@
...
+  }
+
+  if (close_and_return) {
+    return;
+  }
+
+  OpenWebPanel(item);
+}
 
 void SidebarWebPanelController::OpenWebPanel(const SidebarItem& item) {
```
Comment (sangwoo108) [inline browser/ui/sidebar/sidebar_web_panel_controller.cc:55]: How about checking panel_item_ before setting?

```
CHECK(!panel_item_.IsValidItem())
```

File: browser/ui/sidebar/sidebar_web_panel_controller.cc [hunk 9]
```diff
@@ -7,23 +7,48 @@
...
 }
 
-SidebarWebPanelController::~SidebarWebPanelController() = default;
+SidebarWebPanelController::~SidebarWebPanelController() {
+  CHECK(!panel_contents_);
+}
+
+void SidebarWebPanelController::ToggleWebPanel(const SidebarItem& item) {
+  // If panel is for |item|, close and return.
+  // Otherwise, open new panel after closing.
```
Comment (sangwoo108) [inline browser/ui/sidebar/sidebar_web_panel_controller.cc:32]: Maybe noting an edge case this logic is covering would be good

```
When item is same but url has been chaned, closes the current panel and reopens the panel with the new url.
```

**Review by sangwoo108 — APPROVED** [review]
LGTM with nits 👍

File: browser/ui/sidebar/sidebar_web_panel_controller.cc (hunk 7 above)
Comment (simonhong) [reply browser/ui/sidebar/sidebar_web_panel_controller.cc:27]: Added!

File: browser/ui/sidebar/sidebar_web_panel_controller.cc (hunk 9 above)
Comment (simonhong) [reply browser/ui/sidebar/sidebar_web_panel_controller.cc:32]: Added.

File: browser/ui/sidebar/sidebar_web_panel_controller.cc (hunk 8 above)
Comment (simonhong) [reply browser/ui/sidebar/sidebar_web_panel_controller.cc:55]: Added.

File: browser/ui/sidebar/sidebar_web_panel_controller.h (hunk 6 above)
Comment (simonhong) [reply browser/ui/sidebar/sidebar_web_panel_controller.h]: good suggestion.

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc (hunk 5 above)
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: done.

File: browser/ui/views/frame/brave_browser_view.cc (hunk 4 above)
Comment (simonhong) [reply browser/ui/views/frame/brave_browser_view.cc]: fixed.

File: chromium_src/chrome/browser/ui/views/frame/browser_view.cc (hunk 3 above)
Comment (simonhong) [reply chromium_src/chrome/browser/ui/views/frame/browser_view.cc]:
```
bool tab_change_between_contents_view_and_web_panel =
```

Comment (simonhong) [reply chromium_src/chrome/browser/ui/views/frame/browser_view.cc]: When activation is changed from web panel to split tab, split tab's contents are not changed.
However, active tab could be changed as split view has two tabs.
I think just having `change_tab_contents &= !IsWebPanelContents(new_contents);` seems simplest way.
If we skip when active tab change from panel to tab, we need to have code to know which split tab is active to give focused state. WDYT?

**Comment by simonhong** [comment]
Kindly ping. thanks :)

**Comment by simonhong** [comment]
Got some test failures after rebase. Changed as draft.

**Comment by zeusshomaly** [comment]
> Got some test failures after rebase. Changed as draft.

So wait if this gave a test failure, does this mean you'll need to rework web panel? I am being very patient for web panel, which will be my most used feature of all time :) Wishing you all the best with it, and am hoping development isn't cancelled because of the test failure. :)

**Comment by rebron** [comment]
> > Got some test failures after rebase. Changed as draft.
> 
> So wait if this gave a test failure, does this mean you'll need to rework web panel? I am being very patient for web panel, which will be my most used feature of all time :) Wishing you all the best with it, and am hoping development isn't cancelled because of the test failure. :)

Test failures happen, development doesn't get cancelled because of test failures. Thanks for continuing to be patient, these larger features are complicated and we want to make sure when we release this feature it's durable.

**Comment by zeusshomaly** [comment]
> > > Got some test failures after rebase. Changed as draft.
> > 
> > 
> > So wait if this gave a test failure, does this mean you'll need to rework web panel? I am being very patient for web panel, which will be my most used feature of all time :) Wishing you all the best with it, and am hoping development isn't cancelled because of the test failure. :)
> 
> Test failures happen, development doesn't get cancelled because of test failures. Thanks for continuing to be patient, these larger features are complicated and we want to make sure when we release this feature it's durable.

Of course mate, understandable. And it's my pleasure. I just thought you said on X to me a while back that we'd have something in Nightly in a few weeks, but I assume not. No fuss :)

Where?

https://x.com/rebron/status/1958185258894926205?s=20

**Comment by simonhong** [comment]
When clicking panel's contents, its tab should be activated but it doesn't work like that after rebase.
Not sure why. checking..

**Comment by zeusshomaly** [comment]
> When clicking panel's contents, its tab should be activated but it doesn't work like that after rebase. Not sure why. checking..

Simon, this is random, but it is relating to web panel, I got very sad news, Edge is sadly going to deprecate their version of the web panel which was the same as ours that is still in development. 

image

This change will start soon, I will miss Edge's, but yeah, speaking about this, I really hope you guys can never remove this sidebar web panel feature for good when it comes to Brave. It'll be my top favourite feature that honestly should stay forever.

Apologies, if it sounded bossy, I was just concerned that if Edge's is ending support (Sadly, I didn't sleep a night thinking about it)  then Brave's will be interfered with.

If you also know when you could start making sidebar apps sync across devices, that would be great as well (https://github.com/brave/brave-browser/issues/48126)

**Comment by simonhong** [comment]
> Simon, this is random, but it is relating to web panel, I got very sad news, Edge is sadly going to deprecate their version of the web panel which was the same as ours that is still in development.
> 

It's interesting. curious why it's deprecated. thanks for the info.

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 10]
```diff
@@ -51,14 +52,20 @@ void BraveMultiContentsView::UseContentsContainerViewForWebPanel() {
     contents_container_view_for_web_panel_ = AddChildView(
         std::make_unique<BraveContentsContainerView>(browser_view_));
     contents_container_view_for_web_panel_->SetVisible(false);
+    web_contents_focused_subscriptions_.push_back(
+        contents_container_view_for_web_panel_->contents_view()
+            ->AddWebContentsFocusedCallback(base::BindRepeating(
+                &BraveMultiContentsView::OnWebContentsFocused,
+                base::Unretained(this))));
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_multi_contents_view.cc:60]: It's safe as subscription is owned by this class.

**Comment by simonhong** [comment]
Need to fix one more - sometimes got this reentrancy bug when activating tab. checking.

`[25302:1777815:1120/092248.250358:FATAL:chrome/browser/ui/tabs/tab_strip_model.cc:134] Check failed: !*guard_flag_.`

It seems focus issue. When NTP tab is activated from web panel tab, location bar gets focused but input is not working. typing but nothing entered in omnibox.

**Comment by zeusshomaly** [comment]
> Need to fix one more - sometimes got this reentrancy bug when activating tab. checking.
> 
> `[25302:1777815:1120/092248.250358:FATAL:chrome/browser/ui/tabs/tab_strip_model.cc:134] Check failed: !*guard_flag_.`

Thanks Simon, you did a lot in the past 2 days, proud of you. Do you think after this it'll come to flags as a flag? Or, if not, at least after a few more stuff? because honestly, it could be time you could add it to nightly, and then we do extensive testing, and I can suggest more stuff, or you can look around and add more features that Edge had in it's own sidebar.

Regards.

No hurry, but you're doing god's work mate

Looking very excited to use web panel, thanks so much for making it happen!

**Comment by simonhong** [comment]
> > Need to fix one more - sometimes got this reentrancy bug when activating tab. checking.
> > `[25302:1777815:1120/092248.250358:FATAL:chrome/browser/ui/tabs/tab_strip_model.cc:134] Check failed: !*guard_flag_.`
> 
> Thanks Simon, you did a lot in the past 2 days, proud of you. Do you think after this it'll come to flags as a flag? Or, if not, at least after a few more stuff? because honestly, it could be time you could add it to nightly, and then we do extensive testing, and I can suggest more stuff, or you can look around and add more features that Edge had in it's own sidebar.
> 
> Regards.
> 
> No hurry, but you're doing god's work mate
> 
> Looking very excited to use web panel, thanks so much for making it happen!

After this PR, we still need more.
Ex, this web panel should be coordinated with other panels.
With this PR, web panel and other panel(ex, readinglist) are showing together.
Only one panel(web panel or other builtin panel) should be shown at once.
image

**Comment by zeusshomaly** [comment]
> > > Need to fix one more - sometimes got this reentrancy bug when activating tab. checking.
> > > `[25302:1777815:1120/092248.250358:FATAL:chrome/browser/ui/tabs/tab_strip_model.cc:134] Check failed: !*guard_flag_.`
> > 
> > Thanks Simon, you did a lot in the past 2 days, proud of you. Do you think after this it'll come to flags as a flag? Or, if not, at least after a few more stuff? because honestly, it could be time you could add it to nightly, and then we do extensive testing, and I can suggest more stuff, or you can look around and add more features that Edge had in it's own sidebar.
> > 
> > Regards.
> > 
> > No hurry, but you're doing god's work mate
> > 
> > Looking very excited to use web panel, thanks so much for making it happen!
> 
> After this PR, we still need more.
> Ex, this web panel should be coordinated with other panels.
> With this PR, web panel and other panel(ex, readinglist) are showing together.
> Only one panel(web panel or other builtin panel) should be shown at once.
> image
> 
No problem. I understand :) and just curious, will you start the sidebar bookmarks sync feature too? I think that could be easily done. 

You're doing gods work keep it up.

**Comment by simonhong** [comment]
> No problem. I understand :) and just curious, will you start the sidebar bookmarks sync feature too? I think that could be easily done.
> 

Yup, could take it later :)

**Comment by zeusshomaly** [comment]
> > No problem. I understand :) and just curious, will you start the sidebar bookmarks sync feature too? I think that could be easily done.
> > 
> 
> Yup, could take it later :)
> 

Cheers mate.

**Comment by simonhong** [comment]
Tab focus control works fine for adding a tab or activing different tabs inclusing panel tab.
But, crash happens in some tab closing scenario. Investigating..

```
[43672:43672:1125/121543.860963:FATAL:brave/components/brave_wayback_machine/brave_wayback_machine_tab_helper.cc:77] Check failed: wayback_state_changed_callback_. 
#0 0x7f15a5f39182 base::debug::CollectStackTrace()
#1 0x7f15a5f150d1 base::debug::StackTrace::StackTrace()
#2 0x7f15a5dccfaa logging::LogMessage::Flush()
#3 0x7f15a5dcce71 logging::LogMessage::~LogMessage()
#4 0x7f15a5da8f92 logging::(anonymous namespace)::CheckLogMessage::~CheckLogMessage()
#5 0x7f15a5da8b3b logging::CheckNoreturnError::~CheckNoreturnError()
#6 0x5e31d72d45b7 BraveWaybackMachineTabHelper::SetWaybackStateChangedCallback()
#7 0x5e31d3b7ff14 WaybackMachineStateManager::OnTabStripModelChanged()
#8 0x5e31d420f6d8 TabStripModel::OnChange()
#9 0x5e31d4216989 TabStripModel::SetSelection()
#10 0x5e31d4215df1 TabStripModel::ActivateTabAt()
#11 0x5e31d6a5ebf8 MultiContentsViewDelegateImpl::WebContentsFocused()
#12 0x5e31d3ce6e33 BraveMultiContentsViewDelegateImpl::WebContentsFocused()
#13 0x5e31d3ce6437 BraveMultiContentsView::OnWebContentsFocused()
#14 0x7f159b132cdb base::RepeatingCallback<>::Run()
#15 0x7f159b1309d4 base::internal::CallbackListBase<>::Notify<>()
#16 0x7f159b130a7c views::WebView::OnWebContentsFocused()
#17 0x7f15a245403c content::WebContentsImpl::WebContentsObserverList::NotifyObservers<>()
#18 0x7f15a24661ff content::WebContentsImpl::NotifyWebContentsFocused()
#19 0x7f15a249ccd6 content::WebContentsViewAura::GotFocus()
#20 0x7f15a2223e93 content::RenderViewHostImpl::RenderWidgetGotFocus()
#21 0x7f15a222f258 content::RenderWidgetHostImpl::GotFocus()
#22 0x7f15a224b2d1 content::RenderWidgetHostViewAura::OnWindowFocused()
#23 0x7f159a95e95a wm::FocusController::SetFocusedWindow()
#24 0x7f159a95e338 wm::FocusController::FocusAndActivateWindow()
#25 0x7f159c9cb241 aura::Window::Focus()
#26 0x7f15a2241f49 content::RenderWidgetHostViewAura::Focus()
#27 0x7f15a249b069 content::WebContentsViewAura::Focus()
#28 0x7f15a2464156 content::WebContentsImpl::Focus()
#29 0x7f159afe27f7 views::View::Focus()
#30 0x7f159afa18ad views::FocusManager::SetFocusedViewWithReason()
#31 0x7f159afa0ff2 views::FocusManager::AdvanceFocus()
#32 0x7f159afa2223 views::FocusManager::AdvanceFocusIfNecessary()
#33 0x7f159afdfab5 views::View::SetFocusBehavior()
#34 0x7f159b12f73c views::WebView::UpdateCrashedOverlayView()
#35 0x7f159b12e6d6 views::WebView::SetWebContents()
#36 0x5e31d6a46535 ContentsWebView::SetWebContents()
#37 0x5e31d69ec02f BrowserView::OnTabDetached()
#38 0x5e31d6693d80 Browser::OnTabStripModelChanged()
#39 0x5e31d3aa8d90 BraveBrowser::OnTabStripModelChanged()
#40 0x5e31d420f6d8 TabStripModel::OnChange()
#41 0x5e31d4211cf2 TabStripModel::SendDetachWebContentsNotifications()
#42 0x5e31d421acc0 TabStripModel::CloseTabs()
#43 0x5e31d421b664 TabStripModel::CloseWebContentsAt()
#44 0x7f15a2475570 content::WebContentsImpl::Close()
#45 0x7f15a217ac42 content::RenderFrameHostImpl::ClosePageIgnoringUnloadEvents()

```

**Comment by simonhong** [comment]
It's ready to review again. It's updated to make more robust for tab switching/focusing. PTAL again @sangwoo108 .

File: browser/ui/sidebar/sidebar_web_panel_controller.h [hunk 11]
```diff
@@ -7,32 +7,45 @@
...
   SidebarWebPanelController(const SidebarWebPanelController&) = delete;
   SidebarWebPanelController& operator=(const SidebarWebPanelController&) =
       delete;
 
-  void OpenWebPanel(const SidebarItem& item);
-  void CloseWebPanel();
-  bool IsShowingWebPanel() const;
+  void ToggleWebPanel(const SidebarItem& item);
+
+  content::WebContents* panel_contents() const { return panel_contents_.get(); }
```
Comment (sangwoo108) [inline browser/ui/sidebar/sidebar_web_panel_controller.h]: I think returning `const content::WebContents*` is right in terms of method constness. If we need non-const web contents, maybe we should remove const qualitifer in this method?

File: browser/ui/views/frame/split_view/brave_contents_container_view.h [hunk 12]
```diff
@@ -35,6 +35,11 @@ class BraveContentsContainerView :
   explicit BraveContentsContainerView(BrowserView* browser_view);
   ~BraveContentsContainerView() override;
 
+  void set_for_web_panel(bool for_web_panel) { for_web_panel_ = for_web_panel; }
```
Comment (sangwoo108) [inline browser/ui/views/frame/split_view/brave_contents_container_view.h]: Can we do this in constructor?

```
BraveContentsContainerView(BrowserView* browser_view, bool for_web_panel = false)
```

Then we could make the member variable const indicating that it won't be changed during the lifecycle of this instance.

```
const bool for_web_panel_
```

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 13]
```diff
@@ -129,6 +143,77 @@ void BraveMultiContentsView::ResetResizeArea() {
...
+void BraveMultiContentsView::UpdateContentsBorderAndOverlay() {
+  if (!contents_container_view_for_web_panel_ ||
+      !contents_container_view_for_web_panel_->GetVisible()) {
+    MultiContentsView::UpdateContentsBorderAndOverlay();
+    return;
+  }
+
+  if (!contents_container_view_for_web_panel_->IsActive()) {
+    contents_container_view_for_web_panel_->UpdateBorderAndOverlay(
+        /*is_in_split*/ true, /*is_active*/ false,
```
Comment (sangwoo108) [inline browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: Just out of curiosity, why are we passing |is_in_split| true here? Could we have a better name for the param in the future?

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 14]
```diff
@@ -129,6 +143,77 @@ void BraveMultiContentsView::ResetResizeArea() {
...
 }
 
+void BraveMultiContentsView::UpdateContentsBorderAndOverlay() {
+  if (!contents_container_view_for_web_panel_ ||
+      !contents_container_view_for_web_panel_->GetVisible()) {
+    MultiContentsView::UpdateContentsBorderAndOverlay();
+    return;
+  }
+
+  if (!contents_container_view_for_web_panel_->IsActive()) {
```
Comment (sangwoo108) [inline browser/ui/views/frame/split_view/brave_multi_contents_view.cc:154]: Could we add a comment for readability?

```suggestion
  if (!contents_container_view_for_web_panel_->IsActive()) {
  // Web panel is visible but inactive. Hide border of web panel.
```

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc [hunk 15]
```diff
@@ -129,6 +143,77 @@ void BraveMultiContentsView::ResetResizeArea() {
...
+  if (!GetActiveContentsView()->web_contents()) {
+    return;
+  }
+
+  // When tab is activated from Tab UI, we don't need to ask activate
+  // |web_view|'s contents to TabStripModel again. It's not activated here when
+  // activating tab by clicking contents.
+  if (auto* tab =
+          tabs::TabInterface::MaybeGetFromContents(web_view->web_contents());
+      tab->IsActivated()) {
```
Comment (sangwoo108) [inline browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: Worth checking |tab| validity?

```suggestion
      tab && tab->IsActivated()) {
```

**Review by sangwoo108 — APPROVED** [review]
Still LGTM, left a few optional comments.

File: browser/ui/sidebar/sidebar_web_panel_controller.h (hunk 11 above)
Comment (simonhong) [reply browser/ui/sidebar/sidebar_web_panel_controller.h]: Done.

File: browser/ui/views/frame/split_view/brave_contents_container_view.h (hunk 12 above)
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_contents_container_view.h]: Much better. done!

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc (hunk 14 above)
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_multi_contents_view.cc:154]: Added.

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc (hunk 13 above)
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: Good catch. It's used for drawing border in the initial version of this PR.
Now, we can pass it as `false` for web panel as we have another flag(`for_web_panel_`).
With `for_web_panel_`, container view can handle border visibility by itself.

File: browser/ui/views/frame/split_view/brave_multi_contents_view.cc (hunk 15 above)
Comment (simonhong) [reply browser/ui/views/frame/split_view/brave_multi_contents_view.cc]: Done.

**Review by simonhong — COMMENTED** [review]
@sangwoo108 Thanks for review. all addressed. 👍🏼

**Comment by simonhong** [comment]
Left a comment to get attention from  @brave/sec-team as this PR adds another space that loads external contents.
`WebPanel` feature is for showing web contents in a dedicated panel(called `WebPanel`).
For that, we added third split view that uses only for web panel and created new pinned tab.
Then, that tab's contents is rendered in that area. That contents is always visible in that view regardless of other 
active tab changes. So it's panel. When web panel's contents is focused, all toolbar UI is tied with that dedicated pinned tab.  Also web panel contents can have same level of security with tab contents because panel's contents is also another tab.
User can interact with web panel and can get same level of contents information with normal tab via toolbar.
Previously, I tried to load external contents in side panel but it lacks useful UI/info about that contents.
So, we're going in that way. It's developing under disabled feature flag and still very early stage. Do you think we need security review before merging this PR?

**Comment by diracdeltas** [comment]
> Left a comment to get attention from @brave/sec-team as this PR adds another space that loads external contents.

yes this definitely needs a security review, please file it in brave/reviews. i'm out on leave now but @ShivanKaul can assign.

**Review by goodov — APPROVED** [review]
`chromium_src/patches` lgtm

**Comment by thypon** [comment]
security gtg

**Comment by luissequeira94** [comment]
Hi guys - looks great!!
When can we expect this to reach nightly or release?
I am pending this to move from Opera to Brave (I'd assume not the only one)

**Comment by Yet-another-NPC-guy** [comment]
> Hi guys - looks great!!
> When can we expect this to reach nightly or release?
> I am pending this to move from Opera to Brave (I'd assume not the only one)

I'm waiting for this to move form Edge, since edge has a bad integration with Google search, they removed sidebar and history screenshots, I have no rearason to use it, I still due to more color palette personalization but sidebars on brave can compensate it.
